src/java/org/apache/cassandra/db/counters/CachedCounterLockManager.java

```diff
@@ -32,49 +33,53 @@
  * Note: this implemetation tries to reduce the chance of having two counters lock each other, but as the counters
  * are identified by the hash of the primary key of the row, it is still possible to
  * have some cross-counter contention for counters with different primary keys but the same hash.
- *
+ * <p>
  * This code is copied from
  * <a href="https://github.com/diennea/herddb/blob/master/herddb-utils/src/main/java/herddb/utils/LocalLockManager.java">LocalLockManager</a> from the HerdDB project (Apache 2 licensed).
- *
  */
 public class CachedCounterLockManager implements CounterLockManager
 {
     private final static int EXPECTED_CONCURRENCY = DatabaseDescriptor.getConcurrentCounterWriters() * 16;
-    private final ConcurrentMap<Integer, RefCountedStampedLock> locks = new ConcurrentHashMap<>(EXPECTED_CONCURRENCY);
+    /**
+     * The mapping function in {@link #makeLockForKey(Integer)} relies on the ConcurrentHashMap guarantee that the remapping function is run only once per compute, and that it is run atomically
+     */
+    private final ConcurrentHashMap<Integer, RefCountedLock> locks = new ConcurrentHashMap<>(EXPECTED_CONCURRENCY);
 
     @Override
     public List<LockHandle> grabLocks(Iterable<Integer> keys)
     {
-        List<LockHandle> result = new ArrayList<>();
-        for (Integer key : keys)
-            result.add(makeLockForKey(key));
-        return result;
+        // we must return the locks in order to avoid deadlocks
+        // please note that the list may contain duplicates
+        return StreamSupport.stream(keys.spliterator(), false)
```

Comment:
I believe you told me the other day to avoid streams for performance reasons.
While I don't mind the current implementation please spend 30s thinking if there's something simple we could do to be better, especially considering that the number of keys is typically very small.

Reply:
The alternative is to create some temporary Lists, sort, remove duplicates (create a Set?)... the code will lock uglier and in the end I guess that the amount of temporary objects will be the same,
also in this case IIRC the Stream can optimize some edge cases like empty or one entry collections
I don't have a strong preference here